evm/src/cpu/kernel/asm/ripemd/compression.asm

```diff
@@ -0,0 +1,269 @@
+/// _block is stored in memory and its address virt remains on the stack
+/// Note that STATE takes up 5 stack slots
+/// def compress(state, _block):
+/// 
+///     stateL = state
+///     stateL = loop(stateL)
+///
+///     stateR = state
+///     stateR = loop(stateR)
+///
+///     return mix(state, stateL, stateR)
+///
+///
+/// def mix(stateR, stateL, state):
+///     return
+///         u32(s1 + l2 + r3),
+///         u32(s2 + l3 + r4),
+///         u32(s3 + l4 + r0),
+///         u32(s4 + l0 + r1),
+///         u32(s0 + l1 + r2)
+/// 
+/// where si, li, ri, oi, OF, RD respectively denote 
+/// state[i], stateL[i], stateR[i], output[i], virt, retdest
+
+global compress:
+    // stack:                                       STATE, virt, retdest
+    PUSH switch
+    DUP7
+    %stack () ->     (0, 0, 16, 5, 1)
+    // stack:         0, 0, 16, 5, 1, virt, switch, STATE, virt, retdest
+    DUP12  
+    DUP12  
+    DUP12  
+    DUP12  
+    DUP12
+    // stack:  STATE, 0, 0, 16, 5, 1, virt, switch, STATE, virt, retdest 
+    %jump(loop)
+switch:
+    // stack:                                   STATEL, STATE, virt, retdest
+    PUSH mix
+    DUP12 
+    PUSH 0
+    PUSH 5  
+    PUSH 16 
+    // stack:              16, 5, 0, virt, mix, STATEL, STATE, virt, retdest
+    DUP15
+    DUP15
+    DUP15
+    DUP15
+    DUP15
+    // stack: STATE,       16, 5, 0, virt, mix, STATEL, STATE, virt, retdest
+    %stack (STATE: 5) -> (STATE, 0, 0)
+    // stack: STATE, 0, 0, 16, 5, 0, virt, mix, STATEL, STATE, virt, retdest 
+    %jump(loop)
+mix:
+    // stack: r0, r1, r2, r3, r4, l0, l1, l2, l3, l4, s0, s1, s2, s3, s4, VR, RD 
+    SWAP10
+    // stack: s0, r1, r2, r3, r4, l0, l1, l2, l3, l4, r0, s1, s2, s3, s4, VR, RD 
+    SWAP1
+    // stack: r1, s0, r2, r3, r4, l0, l1, l2, l3, l4, r0, s1, s2, s3, s4, VR, RD 
+    SWAP6
+    // stack: l1, s0, r2, r3, r4, l0, r1, l2, l3, l4, r0, s1, s2, s3, s4, VR, RD 
+    %add3_u32
+    // stack:         o4, r3, r4, l0, r1, l2, l3, l4, r0, s1, s2, s3, s4, VR, RD 
+    SWAP14
```

Comment:
These five `SWAP`s could be one `%stack` macro?

Reply:
unfortunately when I had stack macros that went that deep (I think the algo might be exponential in input prefix depth?), my code wouldn't run

Reply:
Ohhh right

Reply:
Yeah that's fine then!